ql/src/Security/CWE-601/BadRedirectCheck.ql

```diff
@@ -23,35 +23,46 @@ StringOps::HasPrefix checkForLeadingSlash(SsaWithFields v) {
   )
 }
 
-DataFlow::Node checkForSecondSlash(SsaWithFields v) {
-  exists(StringOps::HasPrefix hp | result = hp and hp.getBaseString() = v.getAUse() |
+predicate isCheckedForSecondSlash(SsaWithFields v) {
+  exists(StringOps::HasPrefix hp | hp.getBaseString() = v.getAUse() |
     hp.getSubstring().getStringValue() = "//"
   )
   or
   exists(DataFlow::EqualityTestNode eq, DataFlow::Node slash, DataFlow::ElementReadNode er |
-    result = eq
-  |
     slash.getStringValue() = "/" and
     er.getBase() = v.getAUse() and
     er.getIndex().getIntValue() = 1 and
     eq.eq(_, er, slash)
   )
   or
   // a call to path.Clean will strip away multiple leading slashes
-  exists(DataFlow::CallNode call | call.getTarget().hasQualifiedName("path", "Clean") |
-    call.getArgument(0) = v.getAUse() or
-    call.getASuccessor*() = v.getAUse()
+  isCleaned(v.getAUse())
+}
+
+/**
+ * Holds if `nd` is the result of a call to `path.Clean`, or flows into the first argument
+ * of such a call, possibly inter-procedurally.
+ */
+predicate isCleaned(DataFlow::Node nd) {
+  exists(Function clean | clean.hasQualifiedName("path", "Clean") |
+    nd = clean.getACall()
+    or
+    nd = clean.getACall().getArgument(0)
+  )
+  or
+  isCleaned(nd.getAPredecessor())
```

Comment:
I guess this isn't `forex(DataFlow::Node pred | pred = nd.getAPredecessor() | isCleaned(pred))` because of performance reasons?

Reply:
That's part of it. The other reason is that I'm concerned we might lose cases where there is a check of some sort that ensures the path is clean, and the `Clean` method is only invoked when this check does not hold.

Reply:
In that case, the path is actually clean at the merge point, but we'd not be able to tell since `Clean` was only invoked along one branch.

Reply:
Ah, I see.